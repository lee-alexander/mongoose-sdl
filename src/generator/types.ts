import { DbDefinition, Enum, Schema, SchemaDataType, Model } from '../types';
import { assertUnreachable } from '../util';

export function generateTypes(sdl: DbDefinition): string {
  const enums = Object.entries(sdl.enums).map(([name, data]) => generateEnumType(name, data));
  const models = Object.entries(sdl.models).map(([name, data]) => generateModelType(getModelTypeName(name), data));
  const schemas = Object.entries(sdl.schemas).map(([name, data]) => generateSchemaType(getSchemaTypeName(name), data));

  return [`import { Types, Document } from 'mongoose';`, ...enums, ...models, ...schemas].join('\n\n');
}

function generateEnumType(name: string, data: Enum) {
  const enumValues = data.values.map((value) => `  ${value} = '${value}',`).join('\n');
  return `export enum ${name} {\n${enumValues}\n}`;
}

function generateSchemaType(name: string, data: Schema) {
  const defaultFields = ['  id: string;', '  createdAt: Date;', '  updatedAt: Date;'].join('\n');
  const customFields = Object.entries(data)
    .map(
      ([name, data]) =>
        `  ${data.isImmutable ? 'readonly ' : ''}${name}: ${getTypeName(data.dataType)}${
          data.isRequired ? '' : ' | null | undefined'
        };`
    )
    .join('\n');

  return `export interface ${name} extends Document {\n${defaultFields}\n${customFields}\n}`;
}

function generateModelType(name: string, model: Model) {
  return generateSchemaType(name, model.schema);
}

function getTypeName(field: SchemaDataType): string {
  switch (field.type) {
    case 'String':
      return 'string';
    case 'Boolean':
      return 'boolean';
    case 'Number':
      return 'number';
    case 'Date':
      return 'Date';
    case 'Enum':
      return field.refEnum;
    case 'Schema':
      return getSchemaTypeName(field.refSchema);
    case 'ObjectId':
      return 'Types.ObjectId';
    case 'Array': {
      const elementType = getTypeName(field.elementType);
      return field.elementRequired ? `${elementType}[]` : `(${elementType} | null)[]`;
    }
    case 'Map': {
      const elementType = getTypeName(field.elementType);
      return `Map<string, ${elementType}${field.elementRequired ? '' : ' | null'}>`;
    }
    default:
      assertUnreachable(field);
  }
}

export function getSchemaTypeName(name: string) {
  return `${name}SchemaDocument`;
}

export function getModelTypeName(name: string) {
  return `${name}Document`;
}
