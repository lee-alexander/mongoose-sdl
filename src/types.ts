export interface DbDefinition {
  schemas: {
    [schemaName: string]: Schema;
  };
  enums: {
    [enumName: string]: Enum;
  };
  models: {
    [modelName: string]: Model;
  };
}

export interface Enum {
  values: string[];
}

export interface Model {
  schema: Schema;
}

export interface Schema {
  [fieldName: string]: SchemaField;
}

export interface SchemaField {
  dataType: SchemaDataType;
  isRequired: boolean;
  isIndex: boolean;
  isUnique: boolean;
  isImmutable: boolean;
}

export type SchemaDataType =
  | {
      type: 'String' | 'Number' | 'Boolean' | 'Date';
    }
  | {
      type: 'Enum';
      refEnum: string;
    }
  | {
      type: 'ObjectId';
      refModel: string;
    }
  | {
      type: 'Schema';
      refSchema: string;
    }
  | {
      type: 'Array';
      elementType: SchemaDataTypeWithoutArray;
      elementRequired: boolean;
    };

export type SchemaDataTypeWithoutArray = Exclude<SchemaDataType, { type: 'Array' }>;
