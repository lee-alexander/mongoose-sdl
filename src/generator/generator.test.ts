import { CodegenConfig } from '../types';
import { parseAndGenerate } from '..';

test('generate example.mgsdl', async () => {
  const config: CodegenConfig = {
    inputFile: './sample-data/example.mgsdl',
    outputFile: '',
    externalImportPaths: { UserDetailsInterface: '../externals' },
  };
  const result = await parseAndGenerate(config);
  expect(result).toMatchSnapshot();
});

test('generate topological-ordering.mgsdl', async () => {
  const config: CodegenConfig = {
    inputFile: './sample-data/topological-ordering.mgsdl',
    outputFile: '',
  };
  const result = await parseAndGenerate(config);
  expect(result).toMatchSnapshot();
});

test('generate cycles.mgsdl', async () => {
  const config: CodegenConfig = {
    inputFile: './sample-data/invalid/cycles.mgsdl',
    outputFile: '',
  };
  expect(() => parseAndGenerate(config)).rejects.toMatchInlineSnapshot(`"Cycle detected between schemas"`);
});

test('generate non-virtual-external.mgsdl', async () => {
  const config: CodegenConfig = {
    inputFile: './sample-data/invalid/non-virtual-external.mgsdl',
    outputFile: '',
    externalImportPaths: { UserDetailsInterface: '../externals' },
  };
  expect(() => parseAndGenerate(config)).rejects.toMatchInlineSnapshot(
    `"Cannot use external types for non-virtual fields"`
  );
});

test('generate example.mgsdl without external', async () => {
  const config: CodegenConfig = {
    inputFile: './sample-data/example.mgsdl',
    outputFile: '',
  };
  expect(() => parseAndGenerate(config)).rejects.toMatchInlineSnapshot(
    `"Missing externals in input config: UserDetailsInterface"`
  );
});
