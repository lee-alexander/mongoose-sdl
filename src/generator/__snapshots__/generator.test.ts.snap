// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate example.mgsdl 1`] = `
"import { Types, Document, Schema, model } from 'mongoose';
import { UserDetailsInterface } from '../externals';

export enum UserStatus {
  Pending = 'Pending',
  Active = 'Active',
  Inactive = 'Inactive',
}

export interface UserDocument extends Document {
  id: string;
  createdAt: Date;
  updatedAt: Date;
  email: string;
  name: string | null | undefined;
  status: UserStatus;
  unknownRef: Types.ObjectId;
  fullName: string;
  computedFields: UserDetailsInterface;
}

export interface ProjectDocument extends Document {
  id: string;
  createdAt: Date;
  updatedAt: Date;
  creator: Types.ObjectId;
  details: ProjectDetailsSchemaDocument;
  lastAccessedByUserId: Map<string, string | null>;
  readonly detailsByUserId: Map<string, ProjectDetailsSchemaDocument> | null | undefined;
}

export interface ProjectDetailsSchemaDocument extends Document {
  id: string;
  name: string;
  alternateNames: string[];
  number: number | null | undefined;
  readonly isActive: boolean;
  startDate: Date | null | undefined;
  dynamicRef: Types.ObjectId;
}

export interface MongooseFactoryConfig {
  schemas: {
    User: {
      status: {
        validate: {
          validator: (val: UserStatus) => any,
          message?: string,
        },
      },
      fullName: {
        virtual: {
          get?: (doc: UserSchema) => string
          set?: (doc: UserSchema, value: string) => void
        },
      },
      computedFields: {
        virtual: {
          get?: (doc: UserSchema) => UserDetailsInterface
          set?: (doc: UserSchema, value: UserDetailsInterface) => void
        },
      },
    },
    Project?: {
    },
    ProjectDetails: {
      startDate: {
        validate: {
          validator: (val: Date | null | undefined) => any,
          message?: string,
        },
      },
      dynamicRef: {
        ref: (doc: ProjectDetailsSchema) => 'User' | 'Project',
      },
    }
  }
}

export function initializeMongoose(config: MongooseFactoryConfig) {
  
  const ProjectDetailsSchema = new Schema(
    {
      name: { required: true, type: String },
      alternateNames: { required: true, type: [{ type: String, required: true }] },
      number: { unique: true, type: Number },
      isActive: { required: true, immutable: true, type: Boolean },
      startDate: { type: Date, validate: config.schemas.ProjectDetails.startDate.validate },
      dynamicRef: { required: true, type: Schema.Types.ObjectId, ref: function() { return config.schemas.ProjectDetails.dynamicRef.ref(this); } },
    }
  );
  
  const UserSchema = new Schema(
    {
      email: { required: true, index: true, unique: true, type: String },
      name: { type: String },
      status: { required: true, type: String, enum: UserStatus, validate: config.schemas.User.status.validate },
      unknownRef: { required: true, type: Schema.Types.ObjectId },
    },
    { timestamps: true }
  );
  if (config.schemas.User.fullName.virtual.get) {
    UserSchema.virtual('fullName').get((_, __, doc) => config.schemas.User.virtual.get(doc))
  }
  if (config.schemas.User.fullName.virtual.set) {
    UserSchema.virtual('fullName').set((value, _, doc) => { config.schemas.User.virtual.set(doc, value); })
  }
  if (config.schemas.User.computedFields.virtual.get) {
    UserSchema.virtual('computedFields').get((_, __, doc) => config.schemas.User.virtual.get(doc))
  }
  if (config.schemas.User.computedFields.virtual.set) {
    UserSchema.virtual('computedFields').set((value, _, doc) => { config.schemas.User.virtual.set(doc, value); })
  }
  
  const UserModel = model<UserDocument>('User', UserSchema);
  
  const ProjectSchema = new Schema(
    {
      creator: { required: true, index: true, type: Schema.Types.ObjectId, ref: 'User' },
      details: { required: true, type: ProjectDetailsSchema },
      lastAccessedByUserId: { required: true, type: Map, of: { type: String } },
      detailsByUserId: { immutable: true, type: Map, of: { type: ProjectDetailsSchema, required: true } },
    },
    { timestamps: true }
  );
  
  const ProjectModel = model<ProjectDocument>('Project', ProjectSchema);
  
  return {
    schemas: { UserSchema, ProjectSchema, ProjectDetailsSchema },
    models: { UserModel, ProjectModel }
  };
  
}"
`;

exports[`generate topological-ordering.mgsdl 1`] = `
"import { Types, Document, Schema, model } from 'mongoose';

export interface ProjectDocument extends Document {
  id: string;
  createdAt: Date;
  updatedAt: Date;
  name: string;
  details: ProjectDetailsSchemaDocument;
}

export interface CompanySchemaDocument extends Document {
  id: string;
  name: string | null | undefined;
}

export interface ContractorSchemaDocument extends Document {
  id: string;
  name: string;
  company: CompanySchemaDocument | null | undefined;
  parentContractor: ContractorSchemaDocument;
}

export interface ProjectDetailsSchemaDocument extends Document {
  id: string;
  name: string;
  contractor: ContractorSchemaDocument;
}

export interface MongooseFactoryConfig {
  schemas: {
    Project?: {
    },
    Company?: {
    },
    Contractor?: {
    },
    ProjectDetails?: {
    }
  }
}

export function initializeMongoose(config: MongooseFactoryConfig) {
  
  const CompanySchema = new Schema(
    {
      name: { type: String },
    }
  );
  
  const ContractorSchema = new Schema(
    {
      name: { required: true, type: String },
      company: { type: CompanySchema },
    }
  );
  ContractorSchema.add({ parentContractor: { required: true, type: ContractorSchema } });
  
  const ProjectDetailsSchema = new Schema(
    {
      name: { required: true, type: String },
      contractor: { required: true, type: ContractorSchema },
    }
  );
  
  const ProjectSchema = new Schema(
    {
      name: { required: true, type: String },
      details: { required: true, type: ProjectDetailsSchema },
    },
    { timestamps: true }
  );
  
  const ProjectModel = model<ProjectDocument>('Project', ProjectSchema);
  
  return {
    schemas: { ProjectSchema, CompanySchema, ContractorSchema, ProjectDetailsSchema },
    models: { ProjectModel }
  };
  
}"
`;
